File: C++/Tree/BinaryTree/DiameterOfTree.cpp
#include <iostream>
// #include "Node.cpp"
// #include "TreeBuilder.cpp"
// aleady included in HeightOfTree.cpp
#include "HeightOfTree.cpp"
using namespace std;

// Approach 1
// Time Complexity -> O(n^2)
static int diameterOfTree2(Node *root)
{
    if (root == NULL)
        return 0;
    int dia1 = diameterOfTree(root->left);
    int dia2 = diameterOfTree(root->right);
    int dia3 = height(root->left) + height(root->right) + 1;
    return max(max(dia1, dia2), dia3);
}

// Approach 2
// Time Complexity -> O(n)
static int maxDia = 0;
static int helper(Node *root)
{
    if (root == NULL)
        return 0;
    int leftHeight = helper(root->left);
    int rightHeight = helper(root->right);
    maxDia = max(maxDia, 1 + leftHeight + rightHeight);
    return max(leftHeight, rightHeight) + 1;
}

static int diameterOfTree(Node *root)
{
    helper(root);
    return maxDia;
}

// int main()
// {
//     int nodes[] = {1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1};
//     Node *root = buildTree(nodes);
//     cout << diameterOfTree2(root) << endl;
// }

File: Python/tree/binary_tree/diameter_of_tree.py
from node import Node
from tree_builder import build_tree
from height_of_tree import height


# Approach 1
# Time Complexity -> O(n^2)
def diameter_of_tree2(root: Node) -> int:
    if root is None:
        return 0
    dia_1 = diameter_of_tree2(root.left)
    dia_2 = diameter_of_tree2(root.right)
    dia_3 = height(root.left) + height(root.right) + 1
    return max(dia_1, dia_2, dia_3)


__max_dia = 0


# Approach 2
# Time Complexity -> O(n)
def diameter_of_tree(root: Node) -> int:
    def __helper(root: Node):
        global __max_dia
        if root is None:
            return 0
        left_height = __helper(root.left)
        right_height = __helper(root.right)
        __max_dia = max(__max_dia, left_height + right_height + 1)
        return max(left_height, right_height) + 1

    __helper(root)
    return __max_dia


def main():
    nodes = [1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1]
    root = build_tree(nodes)
    print(diameter_of_tree(root))


if __name__ == "__main__":
    main()
